File: Week08/testbe/src/main/java/com/example/testbe/entities/Employee.java
package com.example.testbe.entities;

import java.time.LocalDate;
import java.time.LocalDateTime;

import com.example.testbe.enums.Gender;

import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Table(name = "employees", uniqueConstraints = @UniqueConstraint(columnNames = "email"))
public class Employee {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Size(min = 4, max = 160)
    @NotBlank
    private String fullName;
    
    @Email
    @NotBlank
    private String email;   

    private LocalDate dateOfBirth;

    @Enumerated(EnumType.STRING)
    private Gender gender;

    @Size(min = 10, max = 10)
    private String phoneNumber;
    private Boolean active;
    private String hashedPassword;
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    @PrePersist
    public void PrePersist() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    public void PreUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

}

File: Week08/testfe/src/types/employee.ts
export interface Employee {
  id: number;
  fullName: string;
  email: string;
  dateOfBirth: string | null;
  gender: "Nam" | "Nu" | "Khac";
  phoneNumber: string;
  active: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface CreateEmployeeRequest {
  fullName: string;
  email: string;
  dateOfBirth?: string | null;
  gender?: "Nam" | "Nu" | "Khac";
  phoneNumber: string;
  active: boolean;
  password: string;
}

export interface UpdateEmployeeRequest {
  fullName: string;
  dateOfBirth?: string | null;
  gender?: "Nam" | "Nu" | "Khac";
  phoneNumber: string;
  active: boolean;
  password?: string | null;
}
